src/main/java/com/linkedin/xinfra/monitor/services/MultiClusterTopicManagementService.java

```diff
@@ -326,32 +338,37 @@ void maybeAddPartitions(int minPartitionNum) throws ExecutionException, Interrup
-            this.getClass().toString(), _topic, partitionNum, minPartitionNum);
-        Set<Integer> blackListedBrokers = _topicFactory.getBlackListedBrokers(_adminClient);
-        Set<BrokerMetadata> brokers = new HashSet<>();
-        for (Node broker : _adminClient.describeCluster().nodes().get()) {
-          BrokerMetadata brokerMetadata = new BrokerMetadata(broker.id(), null);
-          brokers.add(brokerMetadata);
-        }
+      List<TopicPartitionInfo> partitions = topicDescriptions.get(_requestTimeout.toMillis(), TimeUnit.MILLISECONDS).partitions();
 
-        if (!blackListedBrokers.isEmpty()) {
-          brokers.removeIf(broker -> blackListedBrokers.contains(broker.id()));
-        }
+      final int currPartitionNum = partitions.size();
+      if (currPartitionNum >= requiredMinPartitionNum) {
+        LOGGER.debug("{} will not increase partition of the topic {} in the cluster. Current partition count {} and '" +
+                "minimum required partition count is {}.", this.getClass().toString(), _topic, currPartitionNum, requiredMinPartitionNum);
+        return;
+      }
+      LOGGER.info("{} will increase partition of the topic {} in the cluster from {}" + " to {}.",
+              this.getClass().toString(), _topic, currPartitionNum,  requiredMinPartitionNum);
+      Set<BrokerMetadata> brokers = new HashSet<>();
+      for (Node broker : _adminClient.describeCluster().nodes().get(_requestTimeout.toMillis(), TimeUnit.MILLISECONDS)) {
+        BrokerMetadata brokerMetadata = new BrokerMetadata(broker.id(), null);
+        brokers.add(brokerMetadata);
+      }
+      Set<Integer> excludedBrokers = _topicFactory.getExcludedBrokers(_adminClient);
+      if (!excludedBrokers.isEmpty()) {
+        brokers.removeIf(broker ->  excludedBrokers.contains(broker.id()));
+      }
 
-        List<List<Integer>> newPartitionAssignments =
-            newPartitionAssignments(minPartitionNum, partitionNum, brokers, _replicationFactor);
+      List<List<Integer>> newPartitionAssignments =
+              newPartitionAssignments(requiredMinPartitionNum, currPartitionNum, brokers, _replicationFactor);
 
-        NewPartitions newPartitions = NewPartitions.increaseTo(minPartitionNum, newPartitionAssignments);
+      NewPartitions newPartitions = NewPartitions.increaseTo(requiredMinPartitionNum, newPartitionAssignments);
 
-        Map<String, NewPartitions> newPartitionsMap = new HashMap<>();
-        newPartitionsMap.put(_topic, newPartitions);
-        CreatePartitionsResult createPartitionsResult = _adminClient.createPartitions(newPartitionsMap);
-      }
+      Map<String, NewPartitions> newPartitionsMap = new HashMap<>();
+      newPartitionsMap.put(_topic, newPartitions);
+      CreatePartitionsResult createPartitionsResult = _adminClient.createPartitions(newPartitionsMap);
+      createPartitionsResult.all().wait(_requestTimeout.toMillis());
```

Comment:
is it "object.wait(long)"?   use future.get ?

Reply:
Good idea